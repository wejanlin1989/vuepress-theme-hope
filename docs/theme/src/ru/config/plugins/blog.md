---
title: Конфиг плагина блога
icon: blog
order: 2
category:
  - Конфиг
tag:
  - Блог
  - Конфигурация плагина
  - Конфиг темы
---

## Введение

Тема предоставляет функцию блога, и она **не** включена по умолчанию.

Вы можете включить функцию блога, установив для `plugins.blog` значение `true` в настройках темы.

Инструкции смотрите в [Введение в блог](../../guide/blog/intro.md).

## Опции

### autoExcerpt

- Тип: `number | boolean`
- По умолчанию: `false`

Генерировать ли выдержку для каждой страницы. Number value means excerpt length (default is `200`).

::: warning Limitation

Auto excerpt is not available in devServer by default, [see reasons and how to enable it](../theme/basic.md#hotreload).

:::

### filter

- Тип: `(page: Page) => boolean`
- По умолчанию: `(page) => Boolean(page.filePathRelative) && !page.frontmatter.home`

Фильтр страниц, определяющий, следует ли включать страницу.

По умолчанию все страницы, созданные из файлов Markdown, но не домашняя страница, будут включены в качестве статей.

### article

- Тип: `string`
- По умолчанию: `/article/`

Путь маршрута списка статей.

### category

- Тип: `string`
- По умолчанию: `/category/`

Путь маршрута карты категории.

### categoryItem

- Тип: `string`
- По умолчанию: `/category/:name/`

Путь маршрута списка категорий. `:name` будет заменено именем категории.

### tag

- Тип: `string`
- По умолчанию: `/tag/`

Отметьте путь маршрута карты.

### tagItem

- Тип: `string`
- По умолчанию: `/tag/:name/`

Путь маршрута списка тегов. `:name` будет заменено именем тега.

### encrypted

- Тип: `string`
- По умолчанию: `/encrypted/`

Путь к списку зашифрованных статей.

### slide

- Тип: `string`
- По умолчанию: `/slide/`

Путь маршрута списка слайдов.

### star

- Тип: `string`
- По умолчанию: `/star/`

Путь маршрута к списку звездных статей.

### timeline

- Тип: `string`
- По умолчанию: `/timeline/`

Путь маршрута списка временной шкалы.

### hotReload

- Type: `boolean`
- Default: Whether using `--debug` flag

Whether to enable hot reload in the development server.
