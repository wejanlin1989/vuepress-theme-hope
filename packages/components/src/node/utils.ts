import { Logger } from "vuepress-shared/node";

export const logger = new Logger("vuepress-plugin-components");

export const getIconPrefix = (assets = ""): string => {
  if (
    assets === "fontawesome" ||
    assets.match(/^(?:https:)?\/\/kit\.fontawesome\.com\//)
  )
    return "fas fa-";
  if (
    assets === "iconfont" ||
    assets.match(/^(?:https:)?\/\/at\.alicdn\.com\/t\//)
  )
    return "iconfont icon-";

  return "";
};
