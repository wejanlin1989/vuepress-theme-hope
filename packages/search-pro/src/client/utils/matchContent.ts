export type Word = [tag: string, content: string] | string;

const maxLength = 100;
const suffixLength = 20;

export const getMatchedContent = (
  content: string,
  queryString: string
): Word[] | null => {
  const contentLowerCase = content.toLowerCase();
  const queryStringLowerCase = queryString.toLowerCase();
  const result: Word[] = [];

  let startIndex = 0;
  let contentLength = 0;

  const addResult = (content: string, isEnd = false): void => {
    let text = "";

    // a beginning of a long string
    if (contentLength === 0)
      text =
        content.length > suffixLength
          ? `… ${content.slice(-suffixLength)}`
          : content;
    // already the last text
    else if (isEnd)
      text =
        // if the string will be longer than maxLength
        content.length + contentLength > maxLength
          ? `${content.slice(0, maxLength - contentLength)}… `
          : content;
    // text is at the middle
    else
      text =
        content.length > suffixLength
          ? `${content.slice(0, suffixLength)} … ${content.slice(
              -suffixLength
            )}`
          : content;

    if (text) result.push(text);
    contentLength += text.length;

    if (!isEnd) {
      result.push(["strong", queryString]);
      contentLength += queryString.length;

      if (contentLength >= maxLength) result.push(" …");
    }
  };

  let matchIndex = contentLowerCase.indexOf(queryStringLowerCase, startIndex);

  if (matchIndex === -1) return null;

  while (matchIndex >= 0) {
    const endIndex = matchIndex + queryStringLowerCase.length;

    // append content before
    addResult(content.slice(startIndex, matchIndex));

    startIndex = endIndex;

    if (contentLength > maxLength) break;

    matchIndex = contentLowerCase.indexOf(queryStringLowerCase, startIndex);
  }

  if (contentLength < maxLength) addResult(content.slice(startIndex), true);

  return result;
};
