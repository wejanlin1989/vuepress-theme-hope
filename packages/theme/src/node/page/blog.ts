import { isLinkHttp, removeEndingSlash } from "@vuepress/shared";
import { load } from "cheerio";
import matter from "gray-matter";
import {
  HTML_TAGS,
  SVG_TAGS,
  getDate,
  timeTransformer,
} from "vuepress-shared/node";

import { ArticleInfoType, PageType } from "../../shared/index.js";

import type { App, Page } from "@vuepress/core";
import type { AnyNode } from "cheerio";
import type {
  ThemeData,
  ThemePageData,
  ThemeBlogHomePageFrontmatter,
  ThemeProjectHomePageFrontmatter,
  ThemeNormalPageFrontmatter,
} from "../../shared/index.js";

const renderNodes = (app: App, nodes: AnyNode[]): AnyNode[] | [] =>
  Array.isArray(nodes)
    ? nodes
        .map((node) => {
          if (node.type === "tag") {
            // image using relative urls shall be dropped
            if (node.tagName === "img") {
              const { src } = node.attribs;

              // this is a valid image link so we preserve it
              if (isLinkHttp(src) || src.startsWith("/")) return node;

              // The img is probably using alias
              return null;
            }

            if (
              HTML_TAGS.includes(node.tagName) ||
              SVG_TAGS.includes(node.tagName)
            ) {
              // console.log(node.tagName);
              node.children = renderNodes(app, node.children);

              return node;
            }

            // we shall convert `<RouterLink>` to `<a>` tag
            if (node.tagName === "routerlink") {
              console.log("found");

              node.tagName = "a";
              node.attribs["href"] = `${removeEndingSlash(app.options.base)}${
                node.attribs["to"]
              }`;
              node.attribs["target"] = "blank";
              delete node.attribs["to"];
              node.children = renderNodes(app, node.children);

              return node;
            }

            return null;
          }

          return node;
        })
        .filter((node): node is AnyNode => node !== null)
    : nodes;

const $ = load("");

export const getPageExcerpt = (
  app: App,
  page: Page<ThemePageData>,
  autoExcerptLength: number
): string => {
  // get page content
  const content = matter(page.content)
    .content.trim()
    // remove first heading1 as title
    .replace(/^# (.*)$/gm, "");
  let excerpt = "";

  for (const line of content.split("\n")) {
    excerpt += `\n${line}`;
    if (excerpt.length >= autoExcerptLength) break;
  }

  const renderedContent = app.markdown.render(
    excerpt,
    // markdown env
    {
      base: app.options.base,
      filePath: page.filePath,
      filePathRelative: page.filePathRelative,
      frontmatter: { ...page.frontmatter },
    }
  );

  return $.html(renderNodes(app, $.parseHTML(renderedContent)) || []);
};

export const injectBlogInfo = (
  app: App,
  themeData: ThemeData,
  page: Page<ThemePageData>,
  autoExcerpt: number | false,
  injectContentSensitiveData = false
): void => {
  const { config = {} } = themeData.encrypt;
  const frontmatter = page.frontmatter as
    | ThemeProjectHomePageFrontmatter
    | ThemeBlogHomePageFrontmatter
    | ThemeNormalPageFrontmatter;
  const { filePathRelative, path } = page;
  const { createdTime } = page.data.git || {};

  const isArticle =
    // declaring this is an article
    frontmatter.article ||
    // generated from markdown files
    Boolean(frontmatter.article !== false && filePathRelative);
  const isEncrypted = Object.keys(config).some((key) =>
    decodeURI(path).startsWith(key)
  );
  const isSlide = isArticle && frontmatter.layout === "Slide";

  // save page type to routeMeta
  page.routeMeta[ArticleInfoType.type] = frontmatter.home
    ? PageType.home
    : isSlide
    ? PageType.slide
    : isArticle
    ? PageType.article
    : PageType.page;

  // resolve author
  if ("author" in frontmatter)
    page.routeMeta[ArticleInfoType.author] = frontmatter.author;

  // resolve date
  if ("date" in frontmatter) {
    const date = getDate(page.frontmatter.date)?.value;

    if (date) {
      page.routeMeta[ArticleInfoType.date] = frontmatter.date;

      page.routeMeta[ArticleInfoType.localizedDate] = timeTransformer(date, {
        lang: page.lang,
        type: "date",
      });
    }
  } else if (createdTime)
    page.routeMeta[ArticleInfoType.date] = new Date(createdTime);

  if ("category" in frontmatter)
    // resolve category
    // resolve category
    page.routeMeta[ArticleInfoType.category] = frontmatter.category;

  // resolve tag
  if ("tag" in frontmatter)
    page.routeMeta[ArticleInfoType.tag] = frontmatter.tag;

  // resolve sticky
  if ("sticky" in frontmatter)
    page.routeMeta[ArticleInfoType.sticky] = frontmatter.sticky;

  // resolve image
  if ("cover" in frontmatter)
    page.routeMeta[ArticleInfoType.cover] = frontmatter.cover;

  // resolve isOriginal
  if ("isOriginal" in frontmatter)
    page.routeMeta[ArticleInfoType.isOriginal] = frontmatter.isOriginal;

  // resolve encrypted
  if (isEncrypted) page.routeMeta[ArticleInfoType.isEncrypted] = true;

  const excerpt = isEncrypted
    ? ""
    : frontmatter.excerpt ||
      page.excerpt ||
      // special handle auto generated description by seo2 plugin
      (page.data.autoDesc ? "" : frontmatter.description) ||
      // generate excerpt from content
      // excerpt is sensitive with markdown contents
      (injectContentSensitiveData && autoExcerpt
        ? app.markdown.render(
            getPageExcerpt(app, page, autoExcerpt),
            // markdown env
            {
              base: app.options.base,
              filePath: page.filePath,
              filePathRelative: page.filePathRelative,
              frontmatter: { ...page.frontmatter },
            }
          )
        : "");

  // save page excerpt to routeMeta
  if (excerpt) page.routeMeta[ArticleInfoType.excerpt] = excerpt;

  if (
    // reading time data is sensitive with markdown contents
    injectContentSensitiveData &&
    // ensure a valid reading time exists
    page.data.readingTime &&
    page.data.readingTime.words !== 0
  )
    page.routeMeta[ArticleInfoType.readingTime] = page.data.readingTime;
};
