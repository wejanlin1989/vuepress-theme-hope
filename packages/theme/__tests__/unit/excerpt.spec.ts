import { createBaseApp } from "@vuepress/core";
import { path } from "@vuepress/utils";
import { describe, it, expect } from "vitest";

import { emptyTheme } from "./__fixtures__/theme/empty.js";
import { getPageExcerpt } from "../../src/node/page/blog.js";
import type { Page } from "@vuepress/core";
import type { ThemePageData } from "../../src/node/index.js";

describe("Should generate page excerpt correctly", async () => {
  const app = createBaseApp({
    bundler: {} as any,
    source: path.resolve(__dirname, "./__fixtures__/src"),
    theme: emptyTheme,
  });

  await app.init();

  it("Should generate excerpt", () => {
    app.pages.forEach((page) => {
      if (page.path.startsWith("/excerpt/")) {
        expect(
          getPageExcerpt(app, page as Page<ThemePageData>, 200)
        ).toMatchSnapshot();
      }
    });
  });
});
