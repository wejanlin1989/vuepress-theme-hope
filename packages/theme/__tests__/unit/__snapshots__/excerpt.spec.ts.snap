// Vitest Snapshot v1

exports[`Should generate page excerpt correctly > Should generate excerpt 1`] = `
"<p>A text with </p>
"
`;

exports[`Should generate page excerpt correctly > Should generate excerpt 2`] = `
"<p>A long code block which is expected to be cut of</p>
<div class=\\"language-javascript line-numbers-mode\\" data-ext=\\"js\\"><pre v-pre=\\"\\" class=\\"language-javascript\\"><code>const a = 1;
const a = 1;
const a = 1;
const a = 1;
const a = 1;
const a = 1;
const a = 1;
const a = 1;
const a = 1;
const a = 1;
const a = 1;
const a = 1;</code></pre><div class=\\"line-numbers\\" aria-hidden=\\"true\\"><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div></div></div>"
`;

exports[`Should generate page excerpt correctly > Should generate excerpt 3`] = `
"<p>Here is <strong>article excerpt</strong>.</p>
<div class=\\"language-javascript line-numbers-mode\\" data-ext=\\"js\\"><pre v-pre=\\"\\" class=\\"language-javascript\\"><code>const a = 1;
</code></pre><div class=\\"line-numbers\\" aria-hidden=\\"true\\"><div class=\\"line-number\\"></div></div></div><!-- more -->
<h2 id=\\"content\\" tabindex=\\"-1\\"><a class=\\"header-anchor\\" href=\\"#content\\" aria-hidden=\\"true\\">#</a> Content</h2>
<p>Here is main content of <strong>article</strong>.</p>
<ol>
<li>A</li>
<li>B</li>
<li>C</li>
</ol>
<div class=\\"language-javascript line-numbers-mode\\" data-ext=\\"js\\"><pre v-pre=\\"\\" class=\\"language-javascript\\"><code>const a = 1;
</code></pre><div class=\\"line-numbers\\" aria-hidden=\\"true\\"><div class=\\"line-number\\"></div></div></div>"
`;

exports[`Should generate page excerpt correctly > Should generate excerpt 4`] = `
"<h2 id=\\"heading-2\\" tabindex=\\"-1\\"><a class=\\"header-anchor\\" href=\\"#heading-2\\" aria-hidden=\\"true\\">#</a> Heading 2</h2>
<h3 id=\\"heading-3\\" tabindex=\\"-1\\"><a class=\\"header-anchor\\" href=\\"#heading-3\\" aria-hidden=\\"true\\">#</a> Heading 3</h3>
<h4 id=\\"heading-4\\" tabindex=\\"-1\\"><a class=\\"header-anchor\\" href=\\"#heading-4\\" aria-hidden=\\"true\\">#</a> Heading 4</h4>
<h5 id=\\"heading-5\\" tabindex=\\"-1\\"><a class=\\"header-anchor\\" href=\\"#heading-5\\" aria-hidden=\\"true\\">#</a> Heading 5</h5>
<h6 id=\\"heading-6\\" tabindex=\\"-1\\"><a class=\\"header-anchor\\" href=\\"#heading-6\\" aria-hidden=\\"true\\">#</a> Heading 6</h6>
<h2 id=\\"text\\" tabindex=\\"-1\\"><a class=\\"header-anchor\\" href=\\"#text\\" aria-hidden=\\"true\\">#</a> Text</h2>
<p>This sentence has <strong>bold</strong>„ÄÅ<em>italic</em> and <s>delete</s> style text.</p>
<h2 id=\\"paragraph\\" tabindex=\\"-1\\"><a class=\\"header-anchor\\" href=\\"#paragraph\\" aria-hidden=\\"true\\">#</a> Paragraph</h2>
<p>This is a paragraph.</p>
<p>This is another paragraph.</p>
"
`;

exports[`Should generate page excerpt correctly > Should generate excerpt 5`] = `
"<p>This is a <a href=\\"/excerpt/demo.html\\" target=\\"blank\\">link</a> referencing other markdown.</p>
<p><img src=\\"/logo.svg\\" alt=\\"logo\\"></p>
<p><img src=\\"/logo.svg\\" alt=\\"logo\\"></p>
<p><img src=\\"https://exmaple.com/logo.svg\\" alt=\\"logo\\"></p>
<p>A inline image , <img src=\\"/logo.svg\\" alt=\\"logo\\"> and a <img src=\\"https://exmaple.com/logo.svg\\" alt=\\"logo\\">.</p>
"
`;
